mlton/atoms/prim-tycons.fun

```diff
@@ -234,4 +234,64 @@ fun layoutApp (c: t,
       else normal ()
    end
 
+fun layoutFormal (c: t, args: (Layout.t * ({isChar: bool}* BindingStrength.t)) vector) =
+   let
+      local
+         open Layout
+         open BindingStrength
+      in
+         val mayAlign = mayAlign
+         val seq = seq
+         val str = str
+      end
+      val ({isChar}, lay) =
+         case Vector.length args of
+            0 => if equals (c, tuple)
+                    then (({isChar = false}, str "unit"))
+                 else
+                    ({isChar = equals (c, defaultChar ())},
+                     (* always use canonical names, even if synonyms exist *)
+                     if equals (c, real RealSize.R32)
```

Comment:
O.k.  Sometimes we set the print name of type constructors when issuing an error or warning to an alias that makes sense at the point of the error/warning.  It seems that `tensor` does have some match non-exhaustive warnings, though I didn't think that that would cause type constructor names to be set.  In any case, the right thing to do might be to clear the print names of all type constructors after elaboration; that should ensure that these primitive types are always printed with their correct names.